backend/src/main/java/ai/verta/modeldb/lineage/LineageDAORdbImpl.java

```diff
@@ -69,166 +293,304 @@ private void validate(List<LineageEntry> inputList, List<LineageEntry> outputLis
+      BlobHashInCommitFunction blobHashInCommitFunction,
+      CommitInBlobHashFunction commitInBlobHashFunction)
+      throws ModelDBException {
     FindAllInputsOutputs.Response.Builder response = FindAllInputsOutputs.Response.newBuilder();
     try (Session session = ModelDBHibernateUtil.getSessionFactory().openSession()) {
-      for (LineageEntry inputoutput : findAllInputsOutputs.getItemsList()) {
+      List<LineageEntryBatchRequest> itemList = findAllInputsOutputs.getItemsList();
+      for (LineageEntryBatchRequest inputoutput : itemList) {
         validate(inputoutput);
-        final List<LineageEntry> inputs = getInputsByOutput(session, inputoutput);
-        final List<LineageEntry> outputs = getOutputsByInput(session, inputoutput);
-        response.addInputs(LineageEntryBatch.newBuilder().addAllItems(inputs));
-        response.addOutputs(LineageEntryBatch.newBuilder().addAllItems(outputs).build());
+        response
+            .addInputs(
+                LineageEntryBatchResponse.newBuilder()
+                    .addAllItems(
+                        getInputsByOutput(
+                            session,
+                            inputoutput,
+                            blobHashInCommitFunction,
+                            commitInBlobHashFunction))
+                    .build())
+            .addOutputs(
+                LineageEntryBatchResponse.newBuilder()
+                    .addAllItems(
+                        getOutputsByInput(
+                            session,
+                            inputoutput,
+                            blobHashInCommitFunction,
+                            commitInBlobHashFunction))
+                    .build());
       }
     }
     return response.build();
   }
 
-  private void deleteLineage(Session session, LineageEntry input, LineageEntry output) {
-    getExisting(session, input, output).ifPresent(session::delete);
+  private void deleteLineage(
+      Session session,
+      LineageEntry lineageEntry,
+      Map<LineageElement, ConnectionEntity> entityIds,
+      BlobHashInCommitFunction blobHashInCommitFunction)
+      throws ModelDBException {
+    LineageElement key = LineageElement.fromProto(session, lineageEntry, blobHashInCommitFunction);
+    ConnectionEntity connectionEntity = entityIds.get(key);
+    if (connectionEntity != null) {
+      deleteConnectionEntity(session, connectionEntity);
+    }
+    entityIds.keySet().remove(key);
   }
 
-  private void addLineage(Session session, LineageEntry input, LineageEntry output) {
-    saveOrUpdate(session, input, output);
+  private void deleteConnectionEntity(Session session,
+      ConnectionEntity connectionEntity) {
+    List connectionEntities =
+        getConnectionEntities(
+            session,
+            CONNECTION_TYPE_ANY,
+            connectionEntity.getEntityId(),
+            connectionEntity.getEntityType());
+    if (connectionEntities.size() < 2) {
```

Comment:
I don't know what this check is validating. Please add a comment.

Reply:
fixed: validate if current entry is used by more than 1 connection, if not, delete it

Reply:
Yes, we can just skip this check and not remove entities. Removed this code